receiver/prometheusexec/end_to_end_test.go

```diff
@@ -0,0 +1,129 @@
+
+import (
+	"context"
+	"path"
+	"sync"
+	"testing"
+	"time"
+
+	"github.com/stretchr/testify/assert"
+	"github.com/stretchr/testify/require"
+	"go.opentelemetry.io/collector/component/componenttest"
+	"go.opentelemetry.io/collector/config"
+	"go.opentelemetry.io/collector/config/configmodels"
+	"go.opentelemetry.io/collector/consumer/consumerdata"
+	"go.opentelemetry.io/collector/exporter/exportertest"
+	"go.uber.org/zap"
+)
+
+func TestEndToEnd(t *testing.T) {
+	factories, err := config.ExampleComponents()
+	assert.NoError(t, err)
+
+	factory := &Factory{}
+	factories.Receivers[factory.Type()] = factory
+
+	config, err := config.LoadConfigFile(t, path.Join(".", "testdata", "config.yaml"), factories)
+
+	assert.NoError(t, err)
+	assert.NotNil(t, config)
+
+	// Receiver without exec key, error expected
+	errorReceiverConfig := config.Receivers["prometheus_exec"]
+	wrapper := new(zap.NewNop(), errorReceiverConfig.(*Config), nil)
+
+	err = wrapper.Start(context.Background(), nil)
+	if err == nil {
+		t.Errorf("end_to_end_test.go didn't get error, was expecting one since this config has no 'exec' key")
+	}
+
+	// Waitgroup to allow the goroutines to finish for the following end-to-end tests
+	var waitGroup sync.WaitGroup
+	waitGroup.Add(2)
+
+	// Normal test with port defined, expose metrics from fake exporter and make sure they're scraped/received
+	go scrapeTest(t, config.Receivers["prometheus_exec/end_to_end_test/1"], &waitGroup)
+
+	// Normal test with port undefined by user, same as previous test
+	go scrapeTest(t, config.Receivers["prometheus_exec/end_to_end_test/2"], &waitGroup)
+
+	waitGroup.Wait()
+}
+
+// scrapeTest scrapes an endpoint twice, and between each scrape waits the correct amount of time for the subprocess exporter (./testdata/end_to_end_metrics_test/expoter.go)
+// to fail and restart, meaning it verifies two things: the scrape is successful (twice), the process was restarted correctly when failed and the underlying
+// Prometheus receiver was correctly stopped and then restarted. For extra testing the metrics are different every time the subprocess exporter is started
+// And the uniqueness of the metric scraped is verified
+func scrapeTest(t *testing.T, receiverConfig configmodels.Receiver, waitGroup *sync.WaitGroup) {
+	defer waitGroup.Done()
+
+	// Create the wrapper
+	sink := &exportertest.SinkMetricsExporterOld{}
+	wrapper := new(zap.NewNop(), receiverConfig.(*Config), sink)
+
+	// Initiate building the embedded configs and managing the subprocess with Start(), and keep track of current time
+	err := wrapper.Start(context.Background(), componenttest.NewNopHost())
+	if err != nil {
+		t.Errorf("end_to_end_test.go got error = %v", err)
+	}
+	now := time.Now()
+	defer wrapper.Shutdown(context.Background())
+
+	var metrics []consumerdata.MetricsData
+
+	// Make sure a first scrape works by checking for metrics in the test metrics exporter "sink", only return true when there are metrics
+	const waitFor = 30 * time.Second
```

Comment:
Reduced the scrape interval, and can reduce this time. This is just how long it could **possibly** wait for, but if the scrape is done before it will exit this waiting function for sure!